diff --git a/xpcom/base/nscore.h b/xpcom/base/nscore.h
--- a/xpcom/base/nscore.h
+++ b/xpcom/base/nscore.h
@@ -165,6 +165,22 @@
 #endif
 
 #ifdef NS_WIN32
+#if defined(MOZ_DISABLE_XPCOM_STDCALL) && defined(__GNUC__)
+
+#define NS_IMPORT __declspec(dllimport)
+#define NS_IMPORT_(type) __declspec(dllimport) type
+#define NS_EXPORT __declspec(dllexport)
+#define NS_EXPORT_(type) __declspec(dllexport) type
+#define NS_IMETHOD_(type) virtual type
+#define NS_IMETHODIMP_(type) type
+#define NS_METHOD_(type) type
+#define NS_CALLBACK_(_type, _name) _type (* _name)
+#define NS_STDCALL
+#define NS_FROZENCALL __cdecl
+#define NS_EXPORT_STATIC_MEMBER_(type) type
+#define NS_IMPORT_STATIC_MEMBER_(type) type
+
+#else
 
 #define NS_IMPORT __declspec(dllimport)
 #define NS_IMPORT_(type) __declspec(dllimport) type __stdcall
@@ -184,6 +200,8 @@
 
 #define NS_EXPORT_STATIC_MEMBER_(type) type
 #define NS_IMPORT_STATIC_MEMBER_(type) type
+
+#endif
 
 #elif defined(XP_OS2) && defined(__declspec)
 
diff --git a/xpcom/reflect/xptcall/src/md/win32/xptcstubs.cpp b/xpcom/reflect/xptcall/src/md/win32/xptcstubs.cpp
--- a/xpcom/reflect/xptcall/src/md/win32/xptcstubs.cpp
+++ b/xpcom/reflect/xptcall/src/md/win32/xptcstubs.cpp
@@ -44,9 +44,15 @@
 #error "This code is for Win32 only"
 #endif
 
+#if defined(MOZ_DISABLE_XPCOM_STDCALL) && defined(__GNUC__)
+#define STDCALL
+#else
+#define STDCALL __stdcall
+#endif
+
 extern "C" {
 
-static nsresult __stdcall
+static nsresult STDCALL
 PrepareAndDispatch(nsXPTCStubBase* self, PRUint32 methodIndex,
                    PRUint32* args, PRUint32* stackBytesToPop)
 {
@@ -157,7 +163,7 @@
 #else
 
 #define STUB_ENTRY(n) \
-nsresult __stdcall nsXPTCStubBase::Stub##n() \
+nsresult STDCALL nsXPTCStubBase::Stub##n() \
 { \
   PRUint32 *args, stackBytesToPop = 0; \
   nsresult result = 0; \
@@ -174,7 +180,7 @@
 #endif /* __GNUC__ */
 
 #define SENTINEL_ENTRY(n) \
-nsresult __stdcall nsXPTCStubBase::Sentinel##n() \
+nsresult STDCALL nsXPTCStubBase::Sentinel##n() \
 { \
     NS_ASSERTION(0,"nsXPTCStubBase::Sentinel called"); \
     return NS_ERROR_NOT_IMPLEMENTED; \
